### PR TITLE
PINS_DEBUGGING and M43: Read pin states

### PR DESCRIPTION
**Pull request by thinkyhead**
Upon request from a printer builder…
- Add an option to report all pin mappings and states
- Add `M43` command to print a single pin state or all pin states

### CODE REVIEW

**Comment by Bob-the-Kuhn** [comment]
I was thinking of something similar to the M43 watch option to help when positioning end stops.  The differences between my concept and what I see of M43 are:
- toggled the LED when the pin changed state
- could turn off the watch function
- checked every 1/10 a second

It would be very nice if you included the LED toggle function/option in M43.

**Comment by thinkyhead** [comment]
M43 doesn't (yet) have an option to monitor only a single pin state. If that is added, then it makes more sense to include the LED toggle. The way it currently works (checking all pins in a loop) some of the digital pins spontaneously toggle, so the LED would tend to flash a lot, and wouldn't be the most reliable feedback.

Currently this is set up to refresh twice per second. Ten times a second is pretty fast for `M43 W` because, in my testing, some pins toggle spontaneously, so too many messages pile up.

The other thing that occurred to me is that probably / possibly the pins should be put into `INPUT_PULLUP` mode instead of `INPUT` mode. Then, I believe, you can test pins by connecting them to GND, which should be safer than connecting them to 5V.

**Comment by Bob-the-Kuhn** [comment]
Agreed - if you're watching all pins then toggling the LED is worthless.

I'm not aware of any I/O pins that change state just because you read the pin register.  I think you're right to suspect that the pins that change state are all inputs that don't have the pull  up enabled.  Static generated by moving air is enough to move a floating CMOS input back & forth.  Many a board designer has learned that the hard way.

**Comment by thinkyhead** [comment]
> M43 watch option to help when positioning end stops

Hmm. `M43 E` to only test endstops, perhaps… When an LCD is in use, display endstop status on the message line…

File: Marlin/pins.h [hunk 1]
```diff
@@ -230,6 +230,9 @@
 #ifndef HEATER_3_PIN
   #define HEATER_3_PIN -1
 #endif
+#ifndef HEATER_4_PIN
+  #define HEATER_4_PIN -1
+#endif
```
Comment (ghost) [inline Marlin/pins.h:235]: Well, I might be wrong, but now Marlin supports "fifth" extruder and heater? (E0 ... E4, H0 ... H4)

Comment (thinkyhead) [reply Marlin/pins.h:235]: It's creeping in, because of one of the boards we support.

Comment (ghost) [reply Marlin/pins.h:235]: I understand it.

**Pull request merged**